verification/spec/relayer/environment.tla

```diff
@@ -0,0 +1,275 @@
+---------------------------- MODULE environment ----------------------------
+
+EXTENDS Naturals, FiniteSets
+
+CONSTANTS MaxHeight, \* maximal height of all the chains in the system
+          NrChains \* number of chains in the system 
+
+VARIABLES chains, \* a function that assigns to each chainID a chain record 
+          pendingDatagrams \* a function that assigns to each chainID a set of pending datagrams
+          
+vars == <<chains, pendingDatagrams>>
+
+ChainIDs == 1..NrChains
+ClientIDs == 1..(NrChains * NrChains)
+Heights == 1..MaxHeight
+
+nullClientID == 0
+nullHeight == 0
+
+Max(S) == CHOOSE x \in S: \A y \in S: y <= x    
+
+(*** Chains ***
+    A set of chain records. 
+    A chain record contains the following fields:
+    
+    - height : an integer between 1 and MaxHeight. 
+      Stores the current height of the chain.
+    
+    - clientHeights : a function that maps chain identifiers to heights.
+      Stores the height of the clients for all other chains on this chain.
+**************)
+Chains ==    
+    [
+        height : Heights \union {nullHeight},
+        clientHeights : [ChainIDs -> Heights \union {nullHeight}]
+    ]
+    
+(*** Datagrams ***
+    A set of datagrams.
+******************)
+Datagrams ==
+    [type : {"CreateClient"}, clientID : ClientIDs, height : Heights]
+    \union
+    [type : {"ClientUpdate"}, clientID : ClientIDs, height : Heights]    
+
+(***************
+Client ID operators
+***************)
+
+\* given a clientID, translate back to srcChainID and dstChainID
+GetChainIDs(clientID) ==
+    LET srcChainID == (clientID - 1) \div NrChains + 1 IN
+    LET dstChainID == ((clientID - 1) % NrChains) + 1 IN
+    [src |-> srcChainID, dst |-> dstChainID]
+    
+\* get the clientID of the client on the chain onChainID, for the chain forChainID
+GetClientID(onChainID, forChainID) ==
+    (onChainID - 1) * NrChains + forChainID
+    
+\* check if the clientID is a valid client ID w.r.t. chainID    
+ValidClientID(chainID, clientID) ==
+    /\ GetChainIDs(clientID).src = chainID
+    /\ GetChainIDs(clientID).dst \in ChainIDs 
+    
+(***************
+Chain update operators
+***************)    
+
+LightClientUpdate(chainID, chain, datagrams) == 
```

Comment:
This function is pretty long so it's a bit hard to read. Can we refactor it into few smaller functions so it's easier understanding the core logic?